Added e2e tests for multi-zone volume provisioning through VCP

sandeeppsunny opened the pull request:
**What type of PR is this**?
/kind bug

**What this PR does / why we need it**:
Added e2e tests to cover multi-zone volume provisioning through vsphere cloud provider (VCP).
The tests cover the following scenarios :
- Verify that zone labels are set correctly on a dynamically created PV with zones specified in the storage class.
- Verify dynamic PV creation fails when zones are not specified in an environment with no shared datastores among all the nodes.
- Verify dynamic PV creation fails when a zone specified does not have any shared datastores under it in the environment.
- Verify dynamic PV creation with zones work in combination with other storage class parameters such as storage policy, datastore and vSAN capabilities. The tests also also brings up pods and verifies volume attachment with them.
- Verify dynamic PV creation using availability zones fails in combination with other storage class parameters such as storage policy, datastore and vSAN capabilities specifications when any of the former mentioned parameters are incompatible with the rest.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:
These tests are written to cover the changes in the following pull requests
#72687 
#72731

Jenkins automation script source code link:
[https://github.com/sandeeppsunny/k8s-jenkins-automation-scripts/blob/k8s-jenkins-testsetup/jobs-scripts/run-zone-tests.sh](url)
All tests passed successfully on the above mentioned changes
`Ran 19 of 2135 Specs in 1203.660 seconds
SUCCESS! -- 19 Passed | 0 Failed | 0 Pending | 2120 Skipped PASS`



**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

sandeeppsunny committed:
Added e2e tests for multi-zone volume provisioning

sandeeppsunny commented:
/cc @abrarshivani  @SandeepPissay  @divyenpatel

frapposelli commented:
/ok-to-test
/sig vmware

frapposelli commented:
/cc @dougm

sandeeppsunny commented:
/retest

sandeeppsunny commented:
/retest

frapposelli commented:
@sandeeppsunny can you please add yourself to the Kubernetes CLA? @SandeepPissay might be able to help you through the process.

sandeeppsunny commented:
> @sandeeppsunny can you please add yourself to the Kubernetes CLA? @SandeepPissay might be able to help you through the process.

Done.

frapposelli commented:
/lgtm

hoegaarden commented on test/e2e/storage/vsphere/bootstrap.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
I don't think you should change the year of a already existing file, it should show the year the file was first added. Is that correct, @nikhita?

sandeeppsunny replied:
Reverted the copyright year back in all the existing files.

sandeeppsunny commented:
/test pull-kubernetes-e2e-gce-100-performance

vladimirvivien commented:
@sandeeppsunny 
Reduce to two commits
- One for the actual code
- Squash the Bazel and copyright related commits to 1.

vladimirvivien commented:
/lgtm

sandeeppsunny commented:
> @sandeeppsunny
> Reduce to two commits
> 
> * One for the actual code
> * Squash the Bazel and copyright related commits to 1.

Reduced the commits to 2.

sandeeppsunny commented:
> @sandeeppsunny
> Reduce to two commits
> 
> * One for the actual code
> * Squash the Bazel and copyright related commits to 1.

Reduced to two commits.

sandeeppsunny commented:
/test pull-kubernetes-e2e-gce

frapposelli commented:
/lgtm

dougm commented on test/e2e/storage/vsphere/vsphere_utils.go:
```diff
@@ -399,6 +411,38 @@ func verifyVSphereVolumesAccessible(c clientset.Interface, pod *v1.Pod, persiste
+		volumePath := pv.Spec.VsphereVolume.VolumePath
+		// Extract datastoreName from the volume path in the pv spec
+		// For example : "vsanDatastore" is extracted from "[vsanDatastore] 25d8b159-948c-4b73-e499-02001ad1b044/volume.vmdk"
+		datastoreName := volumePath[strings.Index(volumePath, "[")+1 : strings.Index(volumePath, "]")]
```
You can use govmomi's `object.DatastorePath.FromString()` to parse VolumePath.

sandeeppsunny replied:
Thanks for pointing this out. Making this change.

dougm commented on test/e2e/storage/vsphere/nodemapper.go:
```diff
@@ -123,6 +139,144 @@ func (nm *NodeMapper) GenerateNodeMap(vSphereInstances map[string]*VSphere, node
+		}
+
+		var validAncestors []mo.ManagedEntity
+		// Filter out only Datacenter, ClusterComputeResource and HostSystem type objects. These objects will be
```
Why filter here?  The only type I can think of that this would exclude is a `Folder`, which should be a valid object to tag.  Or is there another type that could be excluded here?

sandeeppsunny replied:
The type currently being excluded is Folder. As per the current implementation for zone support in PR #72687, zones are discovered only from Host, Cluster and Datacenter type objects.

dougm commented on test/e2e/storage/vsphere/vsphere_zone_support.go:
```diff
@@ -0,0 +1,378 @@
+
+
+	Testbed description :
+	1. cluster-vsan-1 is tagged with zone-a. So, vsanDatastore inherits zone-a since all the hosts under zone-a have vsanDatastore mounted on them.
```
Is there any automation for creating the testbed categories/tags, tagging resources, etc., or is this done manually?

sandeeppsunny replied:
Yes. The entire testing process including tagging, policy creation etc.. is automated through a jenkins job.

dougm replied:
Can you publish the jenkins job config / script somewhere for folks who don't have access to this jenkins instance?

sandeeppsunny replied:
Updated PR description with the jenkins automation script link.

sandeeppsunny commented:
/test pull-kubernetes-e2e-kops-aws

dougm reviewed (commented):
/lgtm

sandeeppsunny commented:
/test pull-kubernetes-e2e-kops-aws

frapposelli commented:
/test pull-kubernetes-e2e-kops-aws

/approve

frapposelli commented:
/assign @gnufied @rootfs 

PTAL

hoegaarden commented:
Maybe interesting for you all:

`pull-kubernetes-e2e-kops-aws` is currently not working and has been removed from the blocking dashboard and moved to an optional pre-submit:
- https://github.com/kubernetes/test-infra/pull/11018
- https://github.com/kubernetes/test-infra/pull/11079
- https://github.com/kubernetes/kubernetes/issues/73444

frapposelli commented:
thanks @hoegaarden for chiming in 👍, flakiness was killing me 😭 😄

divyenpatel commented on test/e2e/storage/vsphere/vsphere_zone_support.go:
```diff
@@ -0,0 +1,378 @@
+)
+
+const (
+	VsanDatastore1  = "vsanDatastore"
```
Everything looks good to me, except this test requires specific names for datastores, policies and zones on the testbed.

It would be good, if we can read these variables from environment, instead of keeping them hard coded.
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_common.go#L55:6

sandeeppsunny replied:
Makes sense. Fixed the code to read the variables from the environment.

frapposelli commented:
@sandeeppsunny PTAL, it would be ideal to have this merged ASAP as code freeze is rapidly approaching

sandeeppsunny commented:
> @sandeeppsunny PTAL, it would be ideal to have this merged ASAP as code freeze is rapidly approaching

I've addressed divyen's comments and squashed all the existing commits to a single one.

divyenpatel commented:
/lgtm

divyenpatel commented:
/approve

divyenpatel commented:
/test pull-kubernetes-e2e-kops-aws

gnufied commented:
/approve

sandeeppsunny commented:
/retest

sandeeppsunny commented:
/test pull-kubernetes-e2e-kops-aws

sandeeppsunny commented:
/retest

skarthiksrinivas commented:
/retest
